imake: do not skip Linux* vendor definitions

trofi opened the pull request:
Initial build failure is onserved on `staging` branch:

```
$ nix-build -A transfig
...
make[1]: Entering directory '/build/transfig.3.2.4'
Imakefile.c:20:2: Include file Imake.tmpl not found
imake: Exit code 1.
  Stop.
```

This happened after 8675ca0 ("xorg.xorgcffiles: 1.0.6 -> 1.0.7")
bump which exposed an old bug in `imake.patch`. `imake.patch`
dropped by accident definiton of `LinuxRedHat` enum and changed
evaluation of the following code:

```
 #if LinuxDistribution == LinuxRedHat
 ...
 # define ProjectRoot /usr
 #endif
```

References to `/usr` were addded in https://gitlab.freedesktop.org/xorg/util/cf/-/commit/d47131ed97ee491bb883c29ec0b106e8d5acfcd3.

Before this change `LinuxDistribution == LinuxRedHat` evaluated to true.
After this change `LinuxDistribution == LinuxRedHat` evaluates to false.

Mechanically I moved `#if 0` right after enum definitions but still
before system autodetection.

Now `transfig` builds successfully.

Closes: https://github.com/NixOS/nixpkgs/issues/135337
Signed-off-by: Sergei Trofimovich <slyich@gmail.com>

trofi committed:
imake: do not skip Linux* vendor definitions

Initial build failure is onserved on `staging` branch:

```
$ nix-build -A transfig
...
make[1]: Entering directory '/build/transfig.3.2.4'
Imakefile.c:20:2: Include file Imake.tmpl not found
imake: Exit code 1.
  Stop.
```

This happened after 8675ca0 ("xorg.xorgcffiles: 1.0.6 -> 1.0.7")
bump which exposed an old bug in `imake.patch`. `imake.patch`
dropped by accident definiton of `LinuxRedHat` enum and changed
evaluation of the following code:

```
 #if LinuxDistribution == LinuxRedHat
 ...
 # define ProjectRoot /usr
 #endif
```

References to `/usr` were addded in https://gitlab.freedesktop.org/xorg/util/cf/-/commit/d47131ed97ee491bb883c29ec0b106e8d5acfcd3.

Before this change `LinuxDistribution == LinuxRedHat` evaluated to true.
After this change `LinuxDistribution == LinuxRedHat` evaluates to false.

Mechanically I moved `#if 0` right after enum definitions but still
before system autodetection.

Now `transfig` builds successfully.

Closes: https://github.com/NixOS/nixpkgs/issues/135337
Signed-off-by: Sergei Trofimovich <slyich@gmail.com>

r-rmcgibbo commented:
Result of `nixpkgs-review pr 135414` at 480f0e34 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>spark</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/5cb8201548466b49d4b9652664ab5b34">xcruiser</a></li>
    <li><a href="https://gist.github.com/d827def2c7f635496a4b723554a24c5c">xvkbd</a></li>
    <li><a href="https://gist.github.com/2fa806bafbad6baee6727a18e71a47d5">xxkb</a></li>
  </ul>
</details>
<details>
  <summary>52 packages skipped due to time constraints:</summary>
  <ul>
    <li>R</li>
    <li>adapta-gtk-theme</li>
    <li>asciidoc-full</li>
    <li>asciidoc-full-with-plugins</li>
    <li>btrbk</li>
    <li>caffeine-ng</li>
    <li>ccache</li>
    <li>ccacheStdenv</li>
    <li>ccacheWrapper</li>
    <li>chrome-gnome-shell</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>30 packages built successfully:</summary>
  <ul>
    <li>Xaw3d</li>
    <li>cernlib</li>
    <li>gbsplay</li>
    <li>gv</li>
    <li>larswm</li>
    <li>lwm</li>
    <li>mlvwm</li>
    <li>nas</li>
    <li>nx-libs</li>
    <li>oneko</li>
    <li>tightvnc</li>
    <li>tkgate</li>
    <li>transfig</li>
    <li>vncrec</li>
    <li>x11_ssh_askpass</li>
    <li>x11docker</li>
    <li>x2goserver</li>
    <li>x2x</li>
    <li>xautolock</li>
    <li>xbrightness</li>
    <li>xearth</li>
    <li>xfig</li>
    <li>xlife</li>
    <li>xmountains</li>
    <li>xorg.imake</li>
    <li>xskat</li>
    <li>xsok</li>
    <li>xspim</li>
    <li>xxgdb</li>
    <li>xzoom</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

trofi commented:
At least `xvkbd` and `xxkb` build failures are caused by this change. Looking at why.

trofi commented:
Exactly 3 these `nixkpgs` packages override `XAPPLOADDIR`:

```
$ git grep XAPPLOADDIR | cat
pkgs/applications/misc/xcruiser/default.nix:    "XAPPLOADDIR=$(out)/etc/X11/app-defaults"
pkgs/applications/misc/xxkb/default.nix:    "XAPPLOADDIR=${placeholder "out"}/etc/X11/app-defaults"
pkgs/tools/X11/xvkbd/default.nix:    "XAPPLOADDIR=$(out)/etc/X11/app-defaults"
```

trofi commented:
It's a combinations of side-effects of leaking `xorg-cf-files`'s store paths and unmodified `CONFDIR`:

```
Makefile:        ETCX11DIR = /etc/X11
Makefile:          CONFDIR = $(ETCX11DIR)
Makefile:           LIBDIR = /nix/store/jbm58kymjv0v7ydbn87i2fqbzcd9wlza-xorg-cf-files-1.0.7/lib/X11`
```

`ETCX11DIR` or `CONFDIR` should be directly overridable.

`LIBDIR` leaks out of `xorg-cf-files` by hardcoding `ProjectRoot`:

```
site.def.in:#ifndef ProjectRoot
site.def.in:#define ProjectRoot @prefix@

X11.tmpl:#  define LibDir Concat(ProjectRoot,/lib/X11)
```

Let's override `LIBDIR` in leaf files.

trofi committed:
xxkb: explicitly set LIBDIR and CONFDIR

In preparation to fix long-standing `imake` bug #135337 we need
to set `LIBDIR` and `CONFDIR` away from incorrectly embedded
`xorg-cf-files`. Before #135337 the package worked by chance.

Issue: https://github.com/NixOS/nixpkgs/issues/135337

trofi committed:
xcruiser: explicitly set LIBDIR and CONFDIR

In preparation to fix long-standing `imake` bug #135337 we need
to set `LIBDIR` and `CONFDIR` away from incorrectly embedded
`xorg-cf-files`. Before #135337 the package worked by chance.

Issue: https://github.com/NixOS/nixpkgs/issues/135337

trofi committed:
xvkbd: explicitly set LIBDIR and CONFDIR

In preparation to fix long-standing `imake` bug #135337 we need
to set `LIBDIR` and `CONFDIR` away from incorrectly embedded
`xorg-cf-files`. Before #135337 the package worked by chance.

Issue: https://github.com/NixOS/nixpkgs/issues/135337

trofi commented:
Should be good to go. PTAL.

trofi commented:
If it helps understanding the fix a bit more I dumped more context in http://trofi.github.io/posts/228-bisects-all-the-way-down.html

r-burns commented:
Nice detective work :)

The transfig failure is happening in staging-next too, as well as a similar error in vncrec. I rebased your commits onto staging-next and it appears to fix both, so can this target staging-next?

trofi commented:
Sounds good. Rebased against `staging-next`.

r-burns reviewed (approved):
Result of `nixpkgs-review pr 135414` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>spark</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python38Packages.cnvkit</li>
    <li>python39Packages.cnvkit</li>
  </ul>
</details>
<details>
  <summary>82 packages built:</summary>
  <ul>
    <li>R</li>
    <li>Xaw3d</li>
    <li>adapta-gtk-theme</li>
    <li>asciidoc-full</li>
    <li>asciidoc-full-with-plugins</li>
    <li>btrbk</li>
    <li>caffeine-ng</li>
    <li>cernlib</li>
    <li>chrome-gnome-shell</li>
    <li>clevis</li>
    <li>collectd</li>
    <li>collectd-data</li>
    <li>dblatexFull</li>
    <li>diffoscope</li>
    <li>disorderfs</li>
    <li>gbsplay</li>
    <li>gnome.gnome-session (gnome.gnome_session)</li>
    <li>gnome.gnome-shell (gnome.gnome_shell)</li>
    <li>gnome.gnome-terminal (gnome.gnome_terminal)</li>
    <li>gnome.gnome-tweaks (gnome.gnome-tweak-tool)</li>
    <li>gnome.pomodoro</li>
    <li>gnomeExtensions.easyScreenCast</li>
    <li>gnomeExtensions.gsconnect</li>
    <li>gnomeExtensions.night-theme-switcher</li>
    <li>gv</li>
    <li>rep (kakounePlugins.rep)</li>
    <li>larswm</li>
    <li>libvmi</li>
    <li>luksmeta</li>
    <li>lwm</li>
    <li>mlvwm</li>
    <li>nas</li>
    <li>nx-libs</li>
    <li>oneko</li>
    <li>pantheon.elementary-session-settings</li>
    <li>phosh</li>
    <li>postgresql11Packages.plr</li>
    <li>postgresql12Packages.plr</li>
    <li>postgresqlPackages.plr (postgresql13Packages.plr)</li>
    <li>python38Packages.rpy2</li>
    <li>python39Packages.rpy2</li>
    <li>qemu_xen (qemu_xen_4_10)</li>
    <li>qemu_xen-light (qemu_xen_4_10-light)</li>
    <li>qubes-core-vchan-xen</li>
    <li>rstudio</li>
    <li>rstudioWrapper</li>
    <li>sage</li>
    <li>sageWithDoc</li>
    <li>skribilo</li>
    <li>spring</li>
    <li>springLobby</li>
    <li>tang</li>
    <li>tightvnc</li>
    <li>tkgate</li>
    <li>transfig</li>
    <li>udiskie</li>
    <li>vimix-gtk-themes</li>
    <li>vncrec</li>
    <li>whitesur-gtk-theme</li>
    <li>x11_ssh_askpass</li>
    <li>x11docker</li>
    <li>x2goclient</li>
    <li>x2goserver</li>
    <li>x2x</li>
    <li>xautolock</li>
    <li>xbrightness</li>
    <li>xcruiser</li>
    <li>xearth</li>
    <li>xen (xenPackages.xen-vanilla ,xenPackages.xen_4_10-vanilla ,xen_4_10)</li>
    <li>xen-light (xenPackages.xen-light ,xenPackages.xen_4_10-light ,xen_4_10-light)</li>
    <li>xen-slim (xenPackages.xen-slim ,xenPackages.xen_4_10-slim ,xen_4_10-slim)</li>
    <li>xfig</li>
    <li>xlife</li>
    <li>xmountains</li>
    <li>xorg.imake</li>
    <li>xskat</li>
    <li>xsok</li>
    <li>xspim</li>
    <li>xvkbd</li>
    <li>xxgdb</li>
    <li>xxkb</li>
    <li>xzoom</li>
  </ul>
</details>
